internal/restorer/restorer.go

```diff
@@ -285,58 +287,110 @@ func (res *Restorer) Snapshot() *restic.Snapshot {
-			if node.Type != "file" {
-				return nil
-			}
+	var (
+		nchecked uint64
+		work     = make(chan mustCheck, 2*nVerifyWorkers)
+	)
 
-			count++
-			stat, err := os.Stat(target)
-			if err != nil {
-				return err
-			}
-			if int64(node.Size) != stat.Size() {
-				return errors.Errorf("Invalid file size: expected %d got %d", node.Size, stat.Size())
-			}
+	g, ctx := errgroup.WithContext(ctx)
 
-			file, err := os.Open(target)
-			if err != nil {
-				return err
-			}
+	// Traverse tree and send jobs to work.
+	g.Go(func() error {
+		defer close(work)
 
-			offset := int64(0)
-			for _, blobID := range node.Content {
-				blobs, _ := res.repo.Index().Lookup(blobID, restic.DataBlob)
-				length := blobs[0].Length - uint(crypto.Extension)
-				buf := make([]byte, length) // TODO do I want to reuse the buffer somehow?
-				_, err = file.ReadAt(buf, offset)
-				if err != nil {
-					_ = file.Close()
-					return err
+		return res.traverseTree(ctx, dst, string(filepath.Separator), *res.sn.Tree, treeVisitor{
+			enterDir: func(node *restic.Node, target, location string) error { return nil },
+			visitNode: func(node *restic.Node, target, _ string) error {
+				if node.Type != "file" {
+					return nil
 				}
-				if !blobID.Equal(restic.Hash(buf)) {
-					_ = file.Close()
-					return errors.Errorf("Unexpected contents starting at offset %d", offset)
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case work <- mustCheck{node, target}:
+					return nil
+				}
+			},
+			leaveDir: func(node *restic.Node, target, location string) error { return nil },
+		})
+	})
+
+	for i := 0; i < nVerifyWorkers; i++ {
+		g.Go(func() (err error) {
+			var buf []byte
+			for job := range work {
+				buf, err = res.verifyFile(job.path, job.node, buf)
+				if err != nil {
+					break
 				}
-				offset += int64(length)
+				atomic.AddUint64(&nchecked, 1)
 			}
+			return
+		})
+	}
 
-			err = file.Close()
-			if err != nil {
-				return err
+	return int(nchecked), g.Wait()
+}
+
+// Verify that the file target has the contents of node.
+// buf and the first return value are scratch space, passed around for reuse.
+func (res *Restorer) verifyFile(target string, node *restic.Node, buf []byte) ([]byte, error) {
+	f, err := os.Open(target)
+	if err != nil {
+		return buf, err
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	switch {
+	case err != nil:
+		return nil, err
+	case int64(node.Size) != fi.Size():
+		return buf, errors.Errorf("Invalid file size for %s: expected %d, got %d",
+			target, node.Size, fi.Size())
+	}
+
+	var offset int64
+	for _, blobID := range node.Content {
+		blobs, _ := res.repo.Index().Lookup(blobID, restic.DataBlob)
+		if len(blobs) == 0 {
+			return buf, errors.Errorf("Unable to fetch blob %s", blobID)
+		}
+
+		length := blobs[0].Length - uint(crypto.Extension)
+		if length > uint(cap(buf)) {
+			newcap := uint(2 * cap(buf))
+			if newcap < length {
+				newcap = length
 			}
+			buf = make([]byte, newcap)
```

Comment:
Why do you grow `buf` based on the current capacity, why not simply do something like `buf = make([]byte, length*2)`? No big deal, just curious what's the logic behind proposed implementation.

Reply:
I started with doubling in a loop until newcap >= length, moved the check to eliminate the loop, proved to myself that what I had was correct, then stopped thinking about it. I can implement your suggestion if you want. It should re-allocate less often, at the expense of at most twice the peak memory (and much less than that on average).